Random unit test failure: StreamSegmentContainerRegistryTests.testContainerFailureOnStartup

**Problem description**
```
io.pravega.segmentstore.server.store.StreamSegmentContainerRegistryTests > testContainerFailureOnStartup FAILED
   java.lang.AssertionError: Container is registered even if it failed to start. No exception has been thrown.
       at org.junit.Assert.fail(Assert.java:88)
       at io.pravega.test.common.AssertExtensions.assertThrows(AssertExtensions.java:60)
       at io.pravega.segmentstore.server.store.StreamSegmentContainerRegistryTests.testContainerFailureOnStartup(StreamSegmentContainerRegistryTests.java:136)
```

**Problem location**
`StreamSegmentContainerRegistry` and `StreamSegmentContainerRegistryTests`

**Suggestions for an improvement**
Make use of the fix in StreamSegmentContainerRegistry added a few days back for `getContainer`

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerRegistry.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerRegistry.java
@@ -10,9 +10,7 @@
 package io.pravega.segmentstore.server;
 
 import io.pravega.segmentstore.contracts.ContainerNotFoundException;
-
 import java.time.Duration;
-import java.util.Collection;
 import java.util.concurrent.CompletableFuture;
 
 /**
@@ -23,11 +21,6 @@ public interface SegmentContainerRegistry extends AutoCloseable {
      * Gets the number of registered containers.
      */
     int getContainerCount();
-
-    /**
-     * Gets a read-only collection of registered container ids.
-     */
-    Collection<Integer> getRegisteredContainerIds();
 
     /**
      * Gets a reference to the SegmentContainer with given Id.

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java
@@ -21,7 +21,6 @@ import io.pravega.segmentstore.server.SegmentContainerFactory;
 import io.pravega.segmentstore.server.SegmentContainerRegistry;
 import java.time.Duration;
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
@@ -90,15 +89,10 @@ class StreamSegmentContainerRegistry implements SegmentContainerRegistry {
     }
 
     @Override
-    public Collection<Integer> getRegisteredContainerIds() {
-        return this.containers.keySet();
-    }
-
-    @Override
     public SegmentContainer getContainer(int containerId) throws ContainerNotFoundException {
         Exceptions.checkNotClosed(this.closed.get(), this);
         ContainerWithHandle result = this.containers.getOrDefault(containerId, null);
-        if (result == null) {
+        if (result == null || isShutdown(result.container.state())) {
             throw new ContainerNotFoundException(containerId);
         }
 

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java
@@ -124,7 +124,12 @@ public class StreamSegmentContainerRegistryTests extends ThreadPooledTestSuite {
     @Test
     public void testContainerFailureOnStartup() throws Exception {
         final int containerId = 123;
-        TestContainerFactory factory = new TestContainerFactory(new IntentionalException());
+
+        // We insert a ReusableLatch that will allow us to manually delay the TestContainer's shutdown/closing process
+        // so that we have enough time to verify that calling getContainer() on a currently shutting down container will
+        // throw the appropriate exception.
+        ReusableLatch closeReleaseSignal = new ReusableLatch();
+        TestContainerFactory factory = new TestContainerFactory(new IntentionalException(), closeReleaseSignal);
         @Cleanup
         StreamSegmentContainerRegistry registry = new StreamSegmentContainerRegistry(factory, executorService());
 
@@ -134,7 +139,15 @@ public class StreamSegmentContainerRegistryTests extends ThreadPooledTestSuite {
                 ex -> ex instanceof IntentionalException || (ex instanceof IllegalStateException && ex.getCause() instanceof IntentionalException));
 
         AssertExtensions.assertThrows(
-                "Container is registered even if it failed to start.",
+                "Container is registered even if it failed to start (and is currently shut down).",
+                () -> registry.getContainer(containerId),
+                ex -> ex instanceof ContainerNotFoundException);
+
+        // Unblock container closing, which will, in turn, unblock its de-registration.
+        closeReleaseSignal.release();
+
+        AssertExtensions.assertThrows(
+                "Container is registered even if it failed to start (and has been unregistered).",
                 () -> registry.getContainer(containerId),
                 ex -> ex instanceof ContainerNotFoundException);
     }
@@ -220,18 +233,20 @@ public class StreamSegmentContainerRegistryTests extends ThreadPooledTestSuite {
 
     private class TestContainerFactory implements SegmentContainerFactory {
         private final Exception startException;
+        private final ReusableLatch startReleaseSignal;
 
         TestContainerFactory() {
-            this(null);
+            this(null, null);
         }
 
-        TestContainerFactory(Exception startException) {
+        TestContainerFactory(Exception startException, ReusableLatch startReleaseSignal) {
             this.startException = startException;
+            this.startReleaseSignal = startReleaseSignal;
         }
 
         @Override
         public SegmentContainer createStreamSegmentContainer(int containerId) {
-            return new TestContainer(containerId, this.startException);
+            return new TestContainer(containerId, this.startException, this.startReleaseSignal);
         }
     }
 
@@ -242,13 +257,15 @@ public class StreamSegmentContainerRegistryTests extends ThreadPooledTestSuite {
     private class TestContainer extends AbstractService implements SegmentContainer {
         private final int id;
         private final Exception startException;
+        private final ReusableLatch closeReleaseSignal;
         private Exception stopException;
         private final AtomicBoolean closed;
         private ReusableLatch stopSignal;
 
-        TestContainer(int id, Exception startException) {
+        TestContainer(int id, Exception startException, ReusableLatch closeReleaseSignal) {
             this.id = id;
             this.startException = startException;
+            this.closeReleaseSignal = closeReleaseSignal;
             this.closed = new AtomicBoolean();
         }
 
@@ -270,6 +287,11 @@ public class StreamSegmentContainerRegistryTests extends ThreadPooledTestSuite {
         public void close() {
             if (!this.closed.getAndSet(true)) {
                 FutureHelpers.await(ServiceHelpers.stopAsync(this, executorService()));
+                ReusableLatch signal = this.closeReleaseSignal;
+                if (signal != null) {
+                    // Wait until we are told to complete.
+                    signal.awaitUninterruptibly();
+                }
             }
         }
 

